agent/api/task/task.go

```diff
@@ -261,6 +267,118 @@ func (task *Task) volumeName(name string) string {
 	return "ecs-" + task.Family + "-" + task.Version + "-" + name + "-" + utils.RandHex()
 }
 
+// initializeDockerVolumes checks the volume resource in the task to determine if the agent
+// should create the volume before creating the container
+func (task *Task) initializeDockerVolumes(dockerClient dockerapi.DockerClient, ctx context.Context) error {
+	for i, vol := range task.Volumes {
+		// No need to do this for non-docker volume, eg: host bind/empty volume
+		if vol.Type != DockerVolumeType {
+			continue
+		}
+
+		dockerVolume, ok := vol.Volume.(*taskresourcevolume.DockerVolumeConfig)
+		if !ok {
+			return errors.New("task volume: volume configuration does not match the type 'docker'")
+		}
+		// Agent needs to create task-scoped volume
+		if dockerVolume.Scope == taskresourcevolume.TaskScope {
+			err := task.addTaskScopedVolumes(ctx, dockerClient, &task.Volumes[i])
+			if err != nil {
+				return err
+			}
+		} else {
+			// Agent needs to create shared volume if that's auto provisioned
+			err := task.addSharedVolumes(ctx, dockerClient, &task.Volumes[i])
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// addTaskScopedVolumes adds the task scoped volume into task resources and updates container dependency
+func (task *Task) addTaskScopedVolumes(ctx context.Context, dockerClient dockerapi.DockerClient,
+	vol *TaskVolume) error {
+
+	volumeConfig := vol.Volume.(*taskresourcevolume.DockerVolumeConfig)
+	volumeResource, err := taskresourcevolume.NewVolumeResource(
+		ctx,
+		vol.Name,
+		task.volumeName(vol.Name),
+		volumeConfig.Scope, volumeConfig.Autoprovision,
+		volumeConfig.Driver, volumeConfig.DriverOpts,
+		volumeConfig.Labels, dockerClient)
+	if err != nil {
+		return err
+	}
+
+	vol.Volume = &volumeResource.VolumeConfig
+	task.AddResource(resourcetype.DockerVolumeKey, volumeResource)
+	task.updateContainerVolumeDependency(vol.Name)
+	return nil
+}
+
+// addSharedVolumes adds shared volume into task resources and updates container dependency
+func (task *Task) addSharedVolumes(ctx context.Context, dockerClient dockerapi.DockerClient,
```

Comment:
how is shared volume's cleanup handled? we need to treat this differently from task scoped volume whose resourceStatusToTransition specifies cleanup at container cleanup.

Reply:
yes, the shared volume won't be cleaned up [here](https://github.com/aws/amazon-ecs-agent/blob/29015316435f0b640c3477e1a993f88302e1092a/agent/taskresource/volume/dockervolume.go#L284-L289)